### PR TITLE
c/snap-bootstrap: check mount states via initramfsMountStates

### PR DESCRIPTION
**Pull request by pedronis**
mostly for consistency

mock mount states in a structured way in tests

### CODE REVIEW

**Comment by pedronis** [comment]
```
	for _, s := range mst.mountStates {
		if i < s.size() {
			dir, state := s.isMounted(i)
			mst.c.Check(path, Equals, dir)
			return state, nil
		}
		i -= s.size()
	}
```

is still respecting order, it will check wrt position `i`  across the conceptually concatenated mountStates

**Comment by pedronis** [comment]
letting mock the entire structure is probably overkill,  seeing the direction I took in #8552. Need to think a bit.

File: cmd/snap-bootstrap/cmd_initramfs_mounts_test.go [hunk 1]
```diff
@@ -1260,39 +1066,25 @@ func (s *initramfsMountsSuite) TestInitramfsMountsRunModeStep2EnvRefKernelBootst
...
-		return false, fmt.Errorf("unexpected number of calls: %v", n)
-	})
-	defer restore()
+	n := s.mockMountsState(c,
+		mounted{
+			boot.InitramfsUbuntuBootDir,
+			boot.InitramfsUbuntuSeedDir,
+			boot.InitramfsUbuntuDataDir,
+		},
+		mounted{filepath.Join(boot.InitramfsRunMntDir, "base")},
```
Comment (anonymouse64) [inline cmd/snap-bootstrap/cmd_initramfs_mounts_test.go:1085]: same

File: cmd/snap-bootstrap/cmd_initramfs_mounts_test.go [hunk 2]
```diff
@@ -1331,31 +1123,17 @@ func (s *initramfsMountsSuite) TestInitramfsMountsRunModeStep2EnvRefKernelBootst
...
```
Comment (anonymouse64) [inline cmd/snap-bootstrap/cmd_initramfs_mounts_test.go:1134]: same

File: cmd/snap-bootstrap/cmd_initramfs_mounts_test.go [hunk 3]
```diff
@@ -1375,38 +1153,24 @@ func (s *initramfsMountsSuite) TestInitramfsMountsRunModeStep2EnvRefKernelBootst
...
```
Comment (anonymouse64) [inline cmd/snap-bootstrap/cmd_initramfs_mounts_test.go:1171]: same

File: cmd/snap-bootstrap/cmd_initramfs_mounts_test.go [hunk 4]
```diff
@@ -1434,37 +1198,23 @@ func (s *initramfsMountsSuite) TestInitramfsMountsRunModeStep2EnvRefKernelBootst
...
```
Comment (anonymouse64) [inline cmd/snap-bootstrap/cmd_initramfs_mounts_test.go:1215]: same

File: cmd/snap-bootstrap/cmd_initramfs_mounts_test.go [hunk 5]
```diff
@@ -1492,62 +1242,40 @@ func (s *initramfsMountsSuite) TestInitramfsMountsRunModeStep2EnvRefKernelBootst
...
-	})
-	defer restore()
+	n := s.mockMountsState(c,
+		mounted{boot.InitramfsUbuntuSeedDir},
+		notYetMounted{
+			filepath.Join(boot.InitramfsRunMntDir, "base"),
+			filepath.Join(boot.InitramfsRunMntDir, "kernel"),
+			filepath.Join(boot.InitramfsRunMntDir, "snapd"),
+		},
+		notYetMounted{filepath.Join(boot.InitramfsRunMntDir, "ubuntu-data")},
```
Comment (anonymouse64) [inline cmd/snap-bootstrap/cmd_initramfs_mounts_test.go:1273]: same

File: cmd/snap-bootstrap/cmd_initramfs_mounts_test.go [hunk 6]
```diff
@@ -1556,73 +1284,41 @@ func (s *initramfsMountsSuite) TestInitramfsMountsRecoverModeStep2(c *C) {
...
-	})
-	defer restore()
+	n := s.mockMountsState(c,
+		mounted{boot.InitramfsUbuntuSeedDir},
+		mounted{
+			filepath.Join(boot.InitramfsRunMntDir, "base"),
+			filepath.Join(boot.InitramfsRunMntDir, "kernel"),
+			filepath.Join(boot.InitramfsRunMntDir, "snapd"),
+		},
+		mounted{filepath.Join(boot.InitramfsRunMntDir, "ubuntu-data")},
```
Comment (anonymouse64) [inline cmd/snap-bootstrap/cmd_initramfs_mounts_test.go:1296]: same

File: cmd/snap-bootstrap/cmd_initramfs_mounts_test.go [hunk 7]
```diff
@@ -1556,73 +1284,41 @@ func (s *initramfsMountsSuite) TestInitramfsMountsRecoverModeStep2(c *C) {
...
+		mounted{filepath.Join(boot.InitramfsRunMntDir, "host/ubuntu-data")},
```
Comment (anonymouse64) [inline cmd/snap-bootstrap/cmd_initramfs_mounts_test.go:1320]: same

File: cmd/snap-bootstrap/cmd_initramfs_mounts_test.go [hunk 8]
```diff
@@ -1714,24 +1410,20 @@ func mockDevDiskByLabel(c *C) (string, func()) {
...
-	})
-	defer restore()
+	n := s.mockMountsState(c,
+		mounted{boot.InitramfsUbuntuSeedDir},
+		notYetMounted{
+			filepath.Join(boot.InitramfsRunMntDir, "base"),
+			filepath.Join(boot.InitramfsRunMntDir, "kernel"),
+			filepath.Join(boot.InitramfsRunMntDir, "snapd"),
+		},
+		notYetMounted{filepath.Join(boot.InitramfsRunMntDir, "ubuntu-data")},
```
Comment (anonymouse64) [inline cmd/snap-bootstrap/cmd_initramfs_mounts_test.go:1422]: same

**Comment by pedronis** [comment]
@anonymouse64 I tried to address the comments by adding some comments/doing some renames

File: cmd/snap-bootstrap/cmd_initramfs_mounts_test.go [hunk 9]
```diff
@@ -171,57 +171,77 @@ func (s *initramfsMountsSuite) TestInitramfsMountsUnknownMode(c *C) {
...
+// these types represent lists of expected mount directories to be
+// checked with IsMounted with an associated mounted state to simulate
+type (
+	expectedMountDirs interface {
+		size() int
+		// dirAndIsMounted returns the dir expected for the
+		// IsMounted call with relative call number callNum
+		// plus the simulated mounted state
+		dirAndIsMounted(callNum int) (dir string, mounted bool)
+	}
```
Comment (anonymouse64) [inline cmd/snap-bootstrap/cmd_initramfs_mounts_test.go:183]: thank you this is easier to understand now

File: cmd/snap-bootstrap/cmd_initramfs_mounts_test.go [hunk 10]
```diff
@@ -171,57 +171,70 @@ func (s *initramfsMountsSuite) TestInitramfsMountsUnknownMode(c *C) {
...
 		}
 		return false, fmt.Errorf("unexpected number of calls: %v", n)
 	})
-	defer restore()
+	s.AddCleanup(r)
+	return &n
+}
+
+func (s *initramfsMountsSuite) TestInitramfsMountsInstallModeStep1(c *C) {
+	s.mockProcCmdlineContent(c, "snapd_recovery_mode= snapd_recovery_system="+s.sysLabel)
```
[the remark this answers is not in the document]
Comment (anonymouse64) [reply cmd/snap-bootstrap/cmd_initramfs_mounts_test.go]: oh also this is probably fine for a followup BTW

**Pull request merged**